controllers/maintenance_mode_checker.go

```diff
@@ -0,0 +1,121 @@
+/*
+ * maintenance_mode_.go
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2019-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	fdbv1beta2 "github.com/FoundationDB/fdb-kubernetes-operator/api/v1beta2"
+)
+
+// updatePods provides a reconciliation step for recreating pods with new pod
+// specs.
+type maintenanceModeChecker struct{}
+
+const (
+	podUpCheckDelayInSeconds = 60
+)
+
+// reconcile runs the reconciler's work.
+func (maintenanceModeChecker) reconcile(ctx context.Context, r *FoundationDBClusterReconciler, cluster *fdbv1beta2.FoundationDBCluster) *requeue {
+	logger := log.WithValues("namespace", cluster.Namespace, "cluster", cluster.Name, "reconciler", "maintenanceModeChecker")
+
+	adminClient, err := r.getDatabaseClientProvider().GetAdminClient(cluster, r.Client)
+	if err != nil {
+		return &requeue{curError: err}
+	}
+	defer adminClient.Close()
+
+	maintenanceZone, err := adminClient.GetMaintenanceZone()
+	if err != nil {
+		return &requeue{curError: err}
+	}
+	// TODO
+	logger.Info("maintenance mode zone", "zone", maintenanceZone)
+	// Cluster is not in maintenance mode
+	if maintenanceZone == "" {
```

Comment:
I believe that it will return an error if a different fault domain is already in maintenance mode. I agree that it would be best to have this gated with a lock.

Reply:
I can gate it with a lock